src/Compiler/ES/Codegen.idr

```diff
@@ -0,0 +1,713 @@
+jsIntOfChar k s = toInt k $ s <+> ".codePointAt(0)"
+
+-- converts a floating point number to an integer.
+jsIntOfDouble : IntKind -> Doc -> Doc
+jsIntOfDouble k = toInt k . callFun1 "Math.trunc"
+
+jsAnyToString : Doc -> Doc
+jsAnyToString s = "(''+" <+> s <+> ")"
+
+-- converts an integer (`Number` or `BigInt`) to a character
+-- by calling `_truncToChar` from the preamble.
+jsCharOfInt : IntKind -> Doc -> Doc
+jsCharOfInt k = callFun1 (esName "truncToChar") . fromInt k
+
+-- Invokes a function from the preamble to check if an bounded
+-- singed integer is withint bound, and - if that's not the case -
+-- truncate it accordingly.
+-- `isBigInt` reflects whether `int` is a `BigInt` or a `Number`.
+--
+-- Note: We can't determine `isBigInt` from the given number of bits, since
+-- when casting from BigInt (for instance, a `Bits64`) to Number
+-- we need to truncate the BigInt
+-- first, otherwise we might lose precision.
+truncateSigned : (isBigInt : Bool) -> (bits : Int) -> (int : Doc) -> Doc
+truncateSigned isBigInt bits =
+   let add = if isBigInt then "BigInt" else "Int"
+    in callFun1 (esName "trunc" ++ add ++ show bits)
+
+-- like `truncateSigned` but for unsigned integers
+truncateUnsigned : (isBigInt : Bool) -> (bits : Int) -> (int : Doc) -> Doc
+truncateUnsigned isBigInt bits =
+   let add = if isBigInt then "BigInt" else "Int"
+    in callFun1 (esName "truncU" ++ add ++ show bits)
+
+-- invokes an arithmetic operation for a bounded integral value.
+-- this is used to implement `boundedIntOp` and `boundedUIntOp`
+-- where the suffix is set to "s" or "u", respectively.
+boundedOp :  (suffix : String)
+          -> (bits : Int)
+          -> (op : String)
+          -> (lhs : Doc)
+          -> (rhs : Doc)
+          -> Doc
+boundedOp s bits o x y = callFun (fastConcat ["_", o, show bits, s]) [x,y]
+
+-- alias for `boundedOp "s"`
+boundedIntOp : Int -> String -> Doc -> Doc -> Doc
+boundedIntOp = boundedOp "s"
+
+-- alias for `boundedOp "u"`
+boundedUIntOp : Int -> String -> Doc -> Doc -> Doc
+boundedUIntOp = boundedOp "u"
+
+-- generates code for a boolean binop, like `>=`.
+boolOp : (op : String) -> (lhs : Doc) -> (rhs : Doc) -> Doc
+boolOp o lhs rhs = "(" <+> binOp o lhs rhs <+> "?1:0)"
+
+-- convert an Idris constant to its JS representation
+jsConstant : Constant -> Core String
+jsConstant (I i)    = pure $ show i
+jsConstant (I8 i)   = pure $ show i
+jsConstant (I16 i)  = pure $ show i
+jsConstant (I32 i)  = pure $ show i
+jsConstant (I64 i)  = pure $ show i ++ "n"
+jsConstant (BI i)   = pure $ show i ++ "n"
+jsConstant (Str s)  = pure $ jsString s
+jsConstant (Ch c)   = pure . jsString $ singleton c
+jsConstant (Db f)   = pure $ show f
+jsConstant WorldVal = pure $ esName "idrisworld"
+jsConstant (B8 i)   = pure $ show i
+jsConstant (B16 i)  = pure $ show i
+jsConstant (B32 i)  = pure $ show i
+jsConstant (B64 i)  = pure $ show i ++ "n"
+jsConstant ty       = error $ "Unsuported constant " ++ show ty
```

Comment:
You know, I was wondering about this too (this error is already there in the current version of the backend). But the Scheme backends return just a dummy string ("#t") for the type constants, so they don't distinguish between the different `Int` and `Bits` types either. In addition, test `node/casts` contains a pattern match on primitive types and that seems to work just fine. I assume, that pattern matches on even these primitive types are converted to proper matches on nullary type constructors (`Core.CompileExpr.NamedConAlt`), so these constants seem to never make an appearance in the IRs.

But of course I might be mistaken here. Anyway, I'll change this to behave the same as the Scheme backends as there seems to be no strong reason to do otherwise.

Reply:
So, I gave this a try with using only basic features like `hcat` and `vcat` and indentation from the existing pretty printer but no grouping nor aligning of code blocks. Test `node/integers` blows up all memory and fails to build on my system. I'm sure this is related to #1614, as the test's (pretty small) main function results in a pathologically huge JS-function of almost 1000 lines and more than half a million characters.

The minimalistic pretty printer added in this PR deals with this just fine and builds the test in less than two seconds.

Reply:
Strange.